[exporter/awsxray] Add aws sdk http error events to x-ray subsegment and strip prefix `AWS.SDK.` from aws remote service name

pxaws opened the pull request:
**Description:** <Describe what has changed.>
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->

- Convert individual HTTP error events into exceptions within subsegments for AWS SDK spans. 
- Normalize the service name from `awsxray.AWSServiceAttribute` attribute by removing the `AWS.SDK.` prefix (in some aws sdk instrumentation, we have added the prefix to produce metrics with the prefix to clearly indicate the resource). This change ensures that X-Ray backend recognizes standard service names like "DynamoDb", "S3", etc., enabling correct identification of AWS service types.


**Link to tracking Issue:** 
NA

**Testing:** 
Unit tests are added.

**Documentation:**
NA

pxaws committed:
add aws sdk http error events to x-ray subsegment

wangzlei commented on exporter/awsxrayexporter/internal/translator/segment.go:
```diff
@@ -79,6 +79,14 @@ func MakeSegmentDocumentString(span ptrace.Span, resource pcommon.Resource, inde
+func isAwsSdkSpan(span ptrace.Span) bool {
+	attributes := span.Attributes()
+	if rpcSystem, ok := attributes.Get(conventions.AttributeRPCSystem); ok {
+		return rpcSystem.Str() == "aws-api"
```
const aws-api

pxaws replied:
I will define the string as a const

wangzlei commented on exporter/awsxrayexporter/internal/translator/cause.go:
```diff
@@ -34,17 +38,23 @@ func makeCause(span ptrace.Span, attributes map[string]pcommon.Value, resource p
 
 	switch {
-	case hasExceptions:
+	case hasExceptions || hasAwsIndividualHTTPError:
```
Does hasExceptions must be true when hasAwsIndividualHTTPError is true? If not, is it better to move hasAwsIndividualHTTPError to a new case?

jknollmeyer replied:
To my knowledge, they are independent of each other. None, either, or both may be `true`. 

A separate `case` would create different logic though. If the cases are separate, then the code is mutually exclusive.

Current: a span will create X-Ray exceptions for both `exception` and `HTTP request failure` events

If we move to a separate case: a span will only create X-Ray exceptions from `exception` events, unless no `exception events` exist, then the span may create X-Ray exceptions from `HTTP request failure` events

So I am in favor of the current structure

wangzlei commented on exporter/awsxrayexporter/internal/translator/cause.go:
```diff
@@ -76,6 +86,22 @@ func makeCause(span ptrace.Span, attributes map[string]pcommon.Value, resource p
+					timestamp := event.Timestamp().String()
+					strs := []string{errorCode.AsString(), timestamp, errorMessage.Str()}
+					message = strings.Join(strs, "@")
+					segmentID := newSegmentID()
```
Is this segment id meaningless?

jknollmeyer replied:
On line 98 we do use it the ID of the exception, since X-Ray exceptions require an ID. I am not aware of any constraints or validations on the ID (for example, whether it needs to be a segment ID format vs. any unique string)

jknollmeyer committed:
Pull Request review changes for #27232

- Define "aws-api" as a const
- Rename some variables in cause.go to make the intended behavior a little more clear
- PR: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/27232

jknollmeyer commented:
New commit covers the following feedback
- "const aws-api"
It also renames some variables around "hasExceptions" to make the intended behavior more clear

For the other two feedbacks, I have suggested that there is no change needed
- https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/27232#discussion_r1338959269
- https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/27232#discussion_r1338955597

I will follow up wangzlei@ on these

srprash reviewed (approved):
LGTM

jknollmeyer committed:
Change the time used for X-Ray exception events to unix epoch time

- timestamp.String() generates a string like "2023-10-10 01:58:24.675761 +0000 UTC", which is not a standardized format
- Under the hood, it is just calling AsTime().String() https://github.com/open-telemetry/opentelemetry-collector/blob/pdata/v0.66.0/pdata/pcommon/timestamp.go#L36
- Time.String() has the warning "The returned string is meant for debugging; for a stable serialized representation, use t.MarshalText, t.MarshalBinary, or t.Format with an explicit format string." https://pkg.go.dev/time#Time.String

jknollmeyer committed:
Merge branch 'main' into add-aws-http-error-event

jknollmeyer committed:
Change the time used for X-Ray exception events to microsecond precision

- This matches X-Ray guidance for timestamps - "Microsecond reoslution is recommended when available" https://docs.aws.amazon.com/xray/latest/devguide/xray-api-segmentdocuments.html#api-segmentdocuments-fields

jknollmeyer committed:
Merge branch 'main' into add-aws-http-error-event

jknollmeyer committed:
Merge branch 'main' into add-aws-http-error-event

jknollmeyer committed:
Merge branch 'open-telemetry:main' into add-aws-http-error-event

jknollmeyer committed:
Merge branch 'main' into add-aws-http-error-event

jknollmeyer commented:
@astencel-sumo @djaglowski can an authorized user merge the pull request? I am not authorized

jknollmeyer committed:
Merge branch 'main' into add-aws-http-error-event

djaglowski reviewed (approved):
All comments appear to be addressed. I've approved CI and believe we can merge if it passes.
